File: blue/src/colors_compute/managed_oci_backend.py
"""Deployment-owned OCI backend buckets, with retired-state checks before purge."""
import re
from .oci import oci_client, bucket_path, object_path

MARKER = '_colors/backend-owner.json'


def check(value, message):
    if not value:
        raise ValueError(message)


async def managed_oci_backend(opts, action, environment=None, runner=None, coordinator_factory=None):
    mode = opts.get('oci-bucket-mode', 'external')
    check(mode in ('external', 'managed'), 'invalid OCI bucket mode')
    if mode == 'external' or any(opts.get(c+'/dry-run') is True or opts.get(c+'/event') == 'build' for c in ('blue', 'red', 'green')):
        return {'status': 'skipped'}
    bucket, region, namespace, compartment, profile = [opts.get(k) for k in ('oci-bucket', 'oci-region', 'oci-namespace', 'oci-compartment-id', 'profile')]
    check(all(isinstance(v, str) for v in (bucket, region, namespace, compartment, profile)) and re.fullmatch(r'[a-z0-9][a-z0-9-]{1,61}[a-z0-9]', bucket) and re.fullmatch(r'[a-z][a-z0-9-]+', region) and re.fullmatch(r'[a-zA-Z0-9]+', namespace) and compartment.startswith('ocid1.') and re.fullmatch(r'[a-z0-9][a-z0-9_-]{0,62}', profile), 'invalid managed OCI identity')
    request = await oci_client(opts, environment, runner)
    path = bucket_path(opts)
    identity = dict(bucket=bucket, region=region, namespace=namespace, compartment=compartment, profile=profile)
    tags = {'colors-profile': profile, 'colors-purpose': 'managed-backend'}
    observed = await request('GET', path)
    owner, deleting = None, False
    try:
        if observed is None:
            # OCI deliberately conflates denied and absent. A complete compartment
            # listing must confirm absence before create or an idempotent delete.
            page = None
            while True:
                listed = await request('GET', path.rsplit('/', 1)[0], query={'compartmentId': compartment, **({'page': page} if page else {})})
                check(listed is not None and 'conflict' not in listed and isinstance(listed['data'], list), 'OCI bucket absence unconfirmed')
                check(not any(b['name'] == bucket for b in listed['data']), 'OCI bucket absence unconfirmed')
                page = listed['headers'].get('opc-next-page')
                if not page:
                    break
            if action == 'finalize' or any(opts.get(c+'/event') == 'delete' for c in ('blue', 'red', 'green')):
                return {'status': 'absent'}
            check(opts.get('compute-require-existing-state') is not True, 'existing managed backend required')
            observed = await request('POST', path.rsplit('/', 1)[0], dict(name=bucket, compartmentId=compartment, freeformTags=tags, publicAccessType='NoPublicAccess', versioning='Enabled'))
            check(observed and not observed.get('conflict'), 'managed backend creation conflict')
            written = await request('PUT', object_path(opts, MARKER), dict(schema=1, identity=identity, status='active'), headers={'if-none-match': '*', 'content-type': 'application/json'})
            check(written and not written.get('conflict'), 'managed backend ownership conflict')
        metadata = observed['data']
        check(metadata.get('compartmentId') == compartment and all(metadata.get('freeformTags', {}).get(k) == v for k, v in tags.items()), 'managed backend ownership mismatch')
        marker_response = await request('GET', object_path(opts, MARKER))
        marker = marker_response['data'] if marker_response else None
        if marker is None and action == 'finalize' and metadata['freeformTags'].get('colors-phase') == 'deleting':
            marker = dict(schema=1, identity=identity, status='deleting')
        check(isinstance(marker, dict) and marker.get('schema') == 1 and marker.get('identity') == identity and marker.get('status') in ('active', 'deleting'), 'managed backend ownership mismatch')
        if action == 'finalize' and metadata['freeformTags'].get('colors-phase') == 'deleting':
            marker = {**marker, 'status': 'deleting'}
        async def update(body):
            updated = await request('PUT', path, body, headers={'if-match': observed['headers']['etag']})
            check(updated and not updated.get('conflict'), 'managed backend metadata conflict')
        if action == 'bootstrap':
            check(marker['status'] == 'active' and metadata['freeformTags'].get('colors-phase') != 'deleting', 'managed backend deletion in progress')
            if metadata.get('versioning') != 'Enabled' or metadata.get('publicAccessType') != 'NoPublicAccess':
                await update(dict(versioning='Enabled', publicAccessType='NoPublicAccess'))
            return {'status': 'ready', 'bucket': bucket}
        check(opts.get('compute-prevent-destroy') is False, 'managed backend deletion protected')
        async def listing(versions=False):
            items, start = [], None
            while True:
                response = await request('GET', path + ('/objectversions' if versions else '/o'), query={('page' if versions else 'start'): start} if start else {})
                check(response and not response.get('conflict'), 'managed backend listing failed')
                page = response['data']
                check(isinstance(page.get('items' if versions else 'objects'), list), 'managed backend listing malformed')
                items += page['items' if versions else 'objects']
                start = response['headers'].get('opc-next-page') if versions else page.get('nextStartWith')
                if not start:
                    return items
        if marker['status'] == 'active':
            if coordinator_factory:
                owner = coordinator_factory(opts, environment=environment, event_prefix='lifecycle/')
            else:
                from .coordinator import Coordinator
                owner = Coordinator(opts, environment=environment, event_prefix='lifecycle/')
            await owner.acquire()
            check((await owner.snapshot())['document']['status'] == 'retired', 'compute must retire before backend deletion')
            for item in await listing():
                key = item['name']
                if key in (MARKER, profile+'/compute/coordination.json'):
                    continue
                check(key.startswith(profile+'/') and key.endswith('.tfstate'), 'managed backend contains unexpected objects')
                state = (await request('GET', object_path(opts, key)))['data']
                check(state.get('version') == 4 and isinstance(state.get('resources'), list) and not any(r.get('instances') for r in state['resources']), 'managed backend contains live state')
            written = await request('PUT', object_path(opts, MARKER), {**marker, 'status': 'deleting'}, headers={'if-match': marker_response['headers']['etag'], 'content-type': 'application/json'})
            check(written and not written.get('conflict'), 'managed backend ownership conflict')
            deleting = True
        await update({'freeformTags': {**metadata['freeformTags'], 'colors-phase': 'deleting'}})
        for item in sorted(await listing(True), key=lambda i: i['name'] == MARKER):
            check(item.get('versionId'), 'managed backend version missing')
            deleted = await request('DELETE', object_path(opts, item['name']), query={'versionId': item['versionId']})
            check(deleted is None or not deleted.get('conflict'), 'managed backend version conflict')
        deleted = await request('DELETE', path)
        check(deleted is None or not deleted.get('conflict'), 'managed backend bucket deletion conflict')
        return {'status': 'destroyed'}
    finally:
        if owner and not deleting:
            await owner.release()

File: red/src/managed-oci-backend.ts
import {ociClient,bucketPath,objectPath} from './oci.ts';
import {Coordinator} from './coordinator.ts';
import {executeBackendCommand,type BackendRunner} from './backend.ts';
type Map=Record<string,any>;
const MARKER='_colors/backend-owner.json';
function check(value:unknown,message:string):asserts value {if(!value)throw Error(message);}
export async function managedOciBackend(opts:Map,action:string,environment:Map=process.env,runner:BackendRunner=executeBackendCommand,coordinatorFactory?:any){
 const mode=opts['oci-bucket-mode']??'external';check(['external','managed'].includes(mode),'invalid OCI bucket mode');
 if(mode==='external'||['blue','red','green'].some(c=>opts[`${c}/dry-run`]===true||opts[`${c}/event`]==='build'))return {status:'skipped'};
 const bucket=opts['oci-bucket'],region=opts['oci-region'],namespace=opts['oci-namespace'],compartment=opts['oci-compartment-id'],profile=opts.profile;
 check(/^[a-z0-9][a-z0-9-]{1,61}[a-z0-9]$/.test(bucket??'')&&/^[a-z][a-z0-9-]+$/.test(region??'')&&/^[a-zA-Z0-9]+$/.test(namespace??'')&&typeof compartment==='string'&&compartment.startsWith('ocid1.')&&/^[a-z0-9][a-z0-9_-]{0,62}$/.test(profile??''),'invalid managed OCI identity');
 const request=await ociClient(opts,environment,runner),path=bucketPath(opts),collection=path.slice(0,path.lastIndexOf('/')),identity={bucket,region,namespace,compartment,profile},tags={'colors-profile':profile,'colors-purpose':'managed-backend'};
 let observed=await request('GET',path),owner:any,deleting=false;
 try{
  if(observed===null){
   let page:string|undefined;do{const listed=await request('GET',collection,undefined,{compartmentId:compartment,...(page?{page}:{})});check(listed&&Array.isArray(listed.data)&&!listed.data.some((b:Map)=>b.name===bucket),'OCI bucket absence unconfirmed');page=listed.headers['opc-next-page'];}while(page);
   if(action==='finalize'||['blue','red','green'].some(c=>opts[`${c}/event`]==='delete'))return {status:'absent'};
   check(opts['compute-require-existing-state']!==true,'existing managed backend required');
   observed=await request('POST',collection,{name:bucket,compartmentId:compartment,freeformTags:tags,publicAccessType:'NoPublicAccess',versioning:'Enabled'});
   check(observed&&!observed.conflict,'managed backend creation conflict');
   const written=await request('PUT',objectPath(opts,MARKER),{schema:1,identity,status:'active'},{},{'if-none-match':'*','content-type':'application/json'});check(written&&!written.conflict,'managed backend ownership conflict');
  }
  const metadata=observed.data;check(metadata?.compartmentId===compartment&&Object.entries(tags).every(([k,v])=>metadata.freeformTags?.[k]===v),'managed backend ownership mismatch');
  const markerResponse=await request('GET',objectPath(opts,MARKER));let marker=markerResponse?.data;
  if(!marker&&action==='finalize'&&metadata.freeformTags['colors-phase']==='deleting')marker={schema:1,identity,status:'deleting'};
  check(marker?.schema===1&&marker.identity&&Object.keys(marker.identity).length===5&&Object.entries(identity).every(([k,v])=>marker.identity[k]===v)&&['active','deleting'].includes(marker.status),'managed backend ownership mismatch');
  if(action==='finalize'&&metadata.freeformTags['colors-phase']==='deleting')marker={...marker,status:'deleting'};
  const update=async(body:Map)=>{const result=await request('PUT',path,body,{},{'if-match':observed!.headers.etag});check(result&&!result.conflict,'managed backend metadata conflict');};
  if(action==='bootstrap'){
   check(marker.status==='active'&&metadata.freeformTags['colors-phase']!=='deleting','managed backend deletion in progress');
   if(metadata.versioning!=='Enabled'||metadata.publicAccessType!=='NoPublicAccess')await update({versioning:'Enabled',publicAccessType:'NoPublicAccess'});
   return {status:'ready',bucket};
  }
  check(opts['compute-prevent-destroy']===false,'managed backend deletion protected');
  const list=async(versions=false)=>{const items:Map[]=[];let start:string|undefined;do{const response=await request('GET',path+(versions?'/objectversions':'/o'),undefined,start?{[versions?'page':'start']:start}:{});check(response&&!response.conflict,'managed backend listing failed');check(Array.isArray(response.data?.[versions?'items':'objects']),'managed backend listing malformed');items.push(...response.data[versions?'items':'objects']);start=versions?response.headers['opc-next-page']:response.data.nextStartWith;}while(start);return items;};
  if(marker.status==='active'){
   owner=coordinatorFactory?coordinatorFactory(opts,{environment,eventPrefix:'lifecycle/'}):new Coordinator(opts,{environment,eventPrefix:'lifecycle/'});
   await owner.acquire();check((await owner.snapshot()).document.status==='retired','compute must retire before backend deletion');
   for(const item of await list()){
    if(item.name===MARKER||item.name===`${profile}/compute/coordination.json`)continue;
    check(item.name.startsWith(`${profile}/`)&&item.name.endsWith('.tfstate'),'managed backend contains unexpected objects');
    const state=(await request('GET',objectPath(opts,item.name)))?.data;check(state?.version===4&&Array.isArray(state.resources)&&!state.resources.some((r:Map)=>r.instances?.length),'managed backend contains live state');
   }
   const written=await request('PUT',objectPath(opts,MARKER),{...marker,status:'deleting'},{},{'if-match':markerResponse!.headers.etag,'content-type':'application/json'});check(written&&!written.conflict,'managed backend ownership conflict');deleting=true;
  }
  await update({freeformTags:{...metadata.freeformTags,'colors-phase':'deleting'}});
  const versions=await list(true);versions.sort((a,b)=>Number(a.name===MARKER)-Number(b.name===MARKER));
  for(const item of versions){check(item.versionId,'managed backend version missing');const deleted=await request('DELETE',objectPath(opts,item.name),undefined,{versionId:item.versionId});check(!deleted?.conflict,'managed backend version conflict');}
  const deleted=await request('DELETE',path);check(!deleted?.conflict,'managed backend bucket deletion conflict');return {status:'destroyed'};
 }finally{if(owner&&!deleting)await owner.release();}
}
